0016-base-primitive-unification.md

```diff
@@ -0,0 +1,145 @@
+This would enable, for example, primitives to have separate output fields for each output variable of a particular circuit, or to take standard error/confidence intervals out of the metadata and put it right alongside the first moment data.
+
+## Design Proposal
+
+We propose to encourage base primitives, like `BaseSampler` and `BaseEstimator`, to have a `run` method typed as
+
+```python
+class BaseFoo:
+    def run(self, tasks: FooTaskLike | Iterable[FooTaskLike]) -> Job[PrimitiveResult[FooTaskResult]]:
+        """Run one or more input tasks and return a result for each one."""
+```
+
+where `FooTask` is equal to or derives from `Task`, `FooTaskLike` is a union type that is easily coercible into a `FooTask` via the static method `FooTask.coerce`, and where `FooTaskResult` derives from `TaskResult`.
+
+The base containers `PrimitiveResult,`, `Task`, and `TaskResult` are described in the next sections.
+`Job[T]` is any `qiskit.provider.JobV1` whose result method returns type `T`.
+
+Any primitive following the above pattern 
+
+```python
+# instantiate the primitive
+foo = Foo()
+
+# runt he primitive with three tasks
+job = foo.run([task0, task1, task2])
+
+# block for results
+result = job.result()
+
+# get data from second task
+result1 = result[1]
+
+# get particular data from this result (the available fields depend on the primitive type and task,
+# and this example is not proposing any specific names)
+alpha_data = result1.data.alpha
+beta_data = result1.data.beta
+expectation_values = result1.data.expectation_values
+```
+
+## Detailed Design
+
+### Task
+
+We propose the concept of a _task_, which we define as _a single circuit along with auxiliary data required to execute the circuit relative to the primitive in question_. This concept is general enough that it can be used for all primitive types, current and future, where we stress that what the “auxiliary data” is can vary between primitive types. 
+
+```python
+# prototype implementation for the RFC
+@dataclass(frozen=True)
+class Task:
+    circuit: QuantumCircuit
+```
+
+Different primitive types (such as `Estimator`) are intended to subclass this class, adding auxiliary fields as 
+required.
+
+### DataBin
+
+A data bin is a namespace for storing data.
+The fields in the namespace, in general, depend on the task that was executed (not the `type` of the task).
+For example, a `Sampler` will have fields for each output (as defined by the OpenQASM 3 spec, but in Qiskit, you can currently take "output" to mean the names of the classical registers) of the task's circuit.
+The value of each field will store the corresponding data.
+
+All primitives will store their data in `DataBin`s.
+There will not be subclassing to the effect of `SamplerDataBin < DataBin`; look instead to `TaskResult` for such needs.
+
+```python
+# Make a new DataBin class. make_data_bin() is analagous to dataclasses.make_dataclass().
```

Comment:
`dataclasses.make_dataclass` takes a first argument which is a class name. Are you intending these databins to be anonymous classes? Or do you want to add an argument for a class name?

Reply:
I was imagining them as anonomous.